catalog/cataloger_diff.go

```diff
@@ -10,17 +10,30 @@ import (
 	"github.com/treeverse/lakefs/logging"
 )
 
-const diffResultsTableName = "catalog_diff_results"
+type DiffTypeCount struct {
+	DiffType int `db:"diff_type"`
+	Count    int `db:"count"`
+}
+
+const (
+	DiffMaxLimit = 1000
 
-func (c *cataloger) Diff(ctx context.Context, repository string, leftBranch string, rightBranch string) (Differences, error) {
+	diffResultsTableName = "catalog_diff_results"
+)
+
+func (c *cataloger) Diff(ctx context.Context, repository string, leftBranch string, rightBranch string, limit int, after string) (Differences, bool, error) {
 	if err := Validate(ValidateFields{
 		{Name: "repository", IsValid: ValidateRepositoryName(repository)},
 		{Name: "leftBranch", IsValid: ValidateBranchName(leftBranch)},
 		{Name: "rightBranch", IsValid: ValidateBranchName(rightBranch)},
 	}); err != nil {
-		return nil, err
+		return nil, false, err
 	}
-	differences, err := c.db.Transact(func(tx db.Tx) (interface{}, error) {
+
+	if limit <= 0 || limit > DiffMaxLimit {
```

Comment:
a bit confused. I thought diff does not get limit

Reply:
This is the public `Diff` function - it should get a limit for pagination.
The actual diff function `doDiff` doesn't get any limit, we limit the part we read from the temporary table.